### PR TITLE
[SPARK-24329][SQL] Remove comments filtering before parsing of CSV files

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Filtering of comments and whitespace has been performed by uniVocity parser already according to parser settings:
https://github.com/apache/spark/blob/branch-2.3/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala#L178-L180

There is no need to repeat the same before uniVocity parser. In this PR, I propose to remove filtering of whitespaces and comments (call of filterCommentAndEmpty) in the parseIterator method of the UnivocityParser object.

## How was this patch tested?

The changes were tested by CSVSuite

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -196,7 +198,7 @@ class UnivocityParser(
       }
       def getPartialResult(): Option[InternalRow] = {
         try {
-          Some(convert(checkedTokens))
+          convert(checkedTokens).headOption
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: @MaxGekk is this change related?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 2]
```diff
@@ -300,14 +302,11 @@ private[csv] object UnivocityParser {
       lines
     }
 
-    val filteredLines: Iterator[String] =
-      CSVUtils.filterCommentAndEmpty(linesWithoutHeader, options)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I feel sure I put this because Univocity has an issue about this before. Wouldn't we better just keep this just in case? I think we already do such things in Spark side redundantly to make sure.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Seems fixed as the tests pass now. I hit a test failure _if I remember this correctly_.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Yes, it is. I changed returned type of the _convert()_ method from InternalRow to Seq[InternalRow] to catch the cases when uniVocity parser returns nulls (comments and empty lines). As a consequence of that, I have to change this function too because it returns Option which is required by the BadRecordException exception. It is safe because Seq can be either empty or contain only one element. And I though it is better to modify body of _getPartialResult()_ than places where BadRecordException is handled.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Probably, you observed issues in old versions of uniVocity parser as @maropu wrote above. I would propose to remove the filtering till we face to the cases when uniVocity's filter doesn't work as it is expected. So, we would submit an issue for uniVocity and revert the changes back.

> I think we already do such things in Spark side redundantly to make sure in few places.

I looked at another places where we do the same but this is only the place where we do filtering directly before uniVocity.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I mean, for example, we do both Parquet's record-level filter and Spark's filter although they are pushed down.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Also, IMO, even if these kinds of bugs fixed in libraries, we keep workaround for a while in most cases, e.g., https://issues.apache.org/jira/browse/SPARK-24109

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Guys, this is speculation about the bug, it seems to me. If the code was added to fix a bug in previous version fo uniVocity parser, where is the tests for that? For now, we filter CSV rows additionally to uniVocity only because we don't trust to the library as far as I understand @HyukjinKwon . Following to the logic, we should duplicate another functionality of uniVocity in CSV datasource. That looks pretty crazy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: @MaxGekk  Maybe add this in PR description

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: @MaxGekk, It doesn't always mean that we have tests. Because it was there from the first place and I tried to remove it, then the tests were broken. I expected to be broken again but seems passed now. So, I'm just guessing that it's fixed.

Usually we trust but we should be careful if there were some issues found. I don't think we should make this case special. I am not seeing meaningful improvement either.

One nit is, BTW, the purpose of `ignoreLeadingWhiteSpaceInRead` and `ignoreTrailingWhiteSpaceInRead` are basically for trimming the whitespaces in the values not for skipping empty lines.

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: We need strong evidence/test cases to make sure the current uniVocity filtering works.

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I wrote a test in the PR: https://github.com/apache/spark/pull/21394 which is passed on the current implementation but fails on this PR. After this PR, lines with multiple whitespaces are not ignored. To ignore such lines, need to set `ignoreLeadingWhiteSpace` to `true`. See https://github.com/uniVocity/univocity-parsers/blob/v2.6.3/src/main/java/com/univocity/parsers/csv/CsvParser.java#L106-L110

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Oh, so now it actually fixes an issue, right? will take a look soon. BTW, I think you can fold the changes in #21394 into here.
